File: src/ast/SimpleReferenceType.java
package ast;

import java.io.StringWriter;

public class SimpleReferenceType extends ParentNode {
    public Reference Qualifier;

    public SimpleReferenceType(Reference qualifier, BaseNode child) {
        super(NodeType.SimpleReferenceType, child);
        Qualifier = qualifier;
    }

    public void PrintQualifier(StringWriter writer) {
       /* if ((Qualifier & Reference.LValue) != 0) {
            writer.write("&");
        }

        if ((Qualifier & Reference.RValue) != 0) {
            writer.write("&&");
        }*/
    }

    @Override
    public void PrintLeft(StringWriter writer) {
        if (child != null) {
            child.PrintLeft(writer);
        } else if (Qualifier != Reference.None) {
            writer.write(" ");
        }

        PrintQualifier(writer);
    }

    @Override
    public boolean HasRightPart() {
        return child != null && child.HasRightPart();
    }

    @Override
    public void PrintRight(StringWriter writer) {
        if (child != null) {
            child.PrintRight(writer);
        }
    }
}

File: src/de/fearlesstobi/demangler/ast/SimpleReferenceType.kt
package de.fearlesstobi.demangler.ast

import java.io.StringWriter

class SimpleReferenceType(val qualifier: Int, child: BaseNode?) : ParentNode(NodeType.SimpleReferenceType, child) {
    fun PrintQualifier(writer: StringWriter) {
        if (qualifier and Reference.LValue != 0) {
            writer.write("&")
        }
        if (qualifier and Reference.RValue != 0) {
            writer.write("&&")
        }
    }

    public override fun printLeft(writer: StringWriter) {
        if (child != null) {
            child.printLeft(writer)
        } else if (qualifier != Reference.None) {
            writer.write(" ")
        }
        PrintQualifier(writer)
    }

    public override fun hasRightPart(): Boolean {
        return child != null && child.hasRightPart()
    }

    public override fun printRight(writer: StringWriter) {
        child?.printRight(writer)
    }

}